// @legendapp/list/keyboard.js
'use strict';

var React = require('react');
var reactNative = require('react-native');
var reactNativeKeyboardController = require('react-native-keyboard-controller');
var reactNativeReanimated = require('react-native-reanimated');
var reanimated = require('@legendapp/list/reanimated');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespace(React);

// src/integrations/keyboard.tsx

// src/utils/helpers.ts
function isFunction(obj) {
  return typeof obj === "function";
}

// src/hooks/useCombinedRef.ts
var useCombinedRef = (...refs) => {
  const callback = React.useCallback((element) => {
    for (const ref of refs) {
      if (!ref) {
        continue;
      }
      if (isFunction(ref)) {
        ref(element);
      } else {
        ref.current = element;
      }
    }
  }, refs);
  return callback;
};

// src/integrations/keyboard.tsx
var KeyboardAvoidingLegendList = React.forwardRef(function KeyboardAvoidingLegendList2(props, forwardedRef) {
  const {
    contentInset: contentInsetProp,
    horizontal,
    onScroll: onScrollProp,
    safeAreaInsetBottom = 0,
    style: styleProp,
    ...rest
  } = props;
  const styleFlattened = reactNative.StyleSheet.flatten(styleProp);
  const refLegendList = React.useRef(null);
  const combinedRef = useCombinedRef(forwardedRef, refLegendList);
  const isIos = reactNative.Platform.OS === "ios";
  const isAndroid = reactNative.Platform.OS === "android";
  const scrollViewRef = reactNativeReanimated.useAnimatedRef();
  const scrollOffsetY = reactNativeReanimated.useSharedValue(0);
  const animatedOffsetY = reactNativeReanimated.useSharedValue(null);
  const scrollOffsetAtKeyboardStart = reactNativeReanimated.useSharedValue(0);
  const mode = reactNativeReanimated.useSharedValue("idle");
  const keyboardInset = reactNativeReanimated.useSharedValue(0);
  const keyboardHeight = reactNativeReanimated.useSharedValue(0);
  const isOpening = reactNativeReanimated.useSharedValue(false);
  const didInteractive = reactNativeReanimated.useSharedValue(false);
  const isKeyboardOpen = reactNativeReanimated.useSharedValue(false);
  const scrollHandler = reactNativeReanimated.useAnimatedScrollHandler(
    (event) => {
      scrollOffsetY.set(event.contentOffset[horizontal ? "x" : "y"]);
      if (onScrollProp) {
        reactNativeReanimated.runOnJS(onScrollProp)(event);
      }
    },
    [onScrollProp, horizontal]
  );
  const setScrollProcessingEnabled = React.useCallback(
    (enabled) => {
      var _a;
      (_a = refLegendList.current) == null ? void 0 : _a.setScrollProcessingEnabled(enabled);
    },
    [refLegendList]
  );
  reactNativeKeyboardController.useKeyboardHandler(
    // biome-ignore assist/source/useSortedKeys: prefer start/move/end
    {
      onStart: (event) => {
        "worklet";
        mode.set("running");
        if (isKeyboardOpen.get() && event.progress === 1 && event.height > 0) {
          return;
        }
        if (!didInteractive.get()) {
          if (event.height > 0) {
            keyboardHeight.set(event.height - safeAreaInsetBottom);
          }
          isOpening.set(event.progress > 0);
          scrollOffsetAtKeyboardStart.set(scrollOffsetY.get());
          animatedOffsetY.set(scrollOffsetY.get());
          reactNativeReanimated.runOnJS(setScrollProcessingEnabled)(false);
        }
      },
      onInteractive: (event) => {
        "worklet";
        if (mode.get() !== "running") {
          reactNativeReanimated.runOnJS(setScrollProcessingEnabled)(false);
        }
        mode.set("running");
        if (!didInteractive.get()) {
          didInteractive.set(true);
        }
        if (isAndroid && !horizontal) {
          keyboardInset.set(Math.max(0, event.height - safeAreaInsetBottom));
        }
      },
      onMove: (event) => {
        "worklet";
        if (!didInteractive.get()) {
          const vIsOpening = isOpening.get();
          const vKeyboardHeight = keyboardHeight.get();
          const vProgress = vIsOpening ? event.progress : 1 - event.progress;
          const targetOffset = Math.max(
            0,
            scrollOffsetAtKeyboardStart.get() + (vIsOpening ? vKeyboardHeight : -vKeyboardHeight) * vProgress
          );
          scrollOffsetY.set(targetOffset);
          animatedOffsetY.set(targetOffset);
          if (!horizontal) {
            keyboardInset.set(Math.max(0, event.height - safeAreaInsetBottom));
          }
        }
      },
      onEnd: (event) => {
        "worklet";
        const wasInteractive = didInteractive.get();
        const vMode = mode.get();
        mode.set("idle");
        if (vMode === "running") {
          if (!wasInteractive) {
            const vIsOpening = isOpening.get();
            const vKeyboardHeight = keyboardHeight.get();
            const targetOffset = Math.max(
              0,
              scrollOffsetAtKeyboardStart.get() + (vIsOpening ? vKeyboardHeight : -vKeyboardHeight) * (vIsOpening ? event.progress : 1 - event.progress)
            );
            scrollOffsetY.set(targetOffset);
            animatedOffsetY.set(targetOffset);
          }
          reactNativeReanimated.runOnJS(setScrollProcessingEnabled)(true);
          didInteractive.set(false);
          isKeyboardOpen.set(event.height > 0);
          if (!horizontal) {
            const newInset = Math.max(0, event.height - safeAreaInsetBottom);
            if (newInset > 0) {
              keyboardInset.set(newInset);
            } else {
              keyboardInset.set(newInset);
              animatedOffsetY.set(scrollOffsetY.get());
            }
          }
        }
      }
    },
    [scrollViewRef, safeAreaInsetBottom]
  );
  const animatedProps = reactNativeReanimated.useAnimatedProps(() => {
    "worklet";
    var _a, _b, _c, _d;
    const vAnimatedOffsetY = animatedOffsetY.get();
    const baseProps = {
      contentOffset: vAnimatedOffsetY === null ? void 0 : {
        x: 0,
        y: vAnimatedOffsetY
      }
    };
    return isIos ? Object.assign(baseProps, {
      contentInset: {
        bottom: ((_a = contentInsetProp == null ? void 0 : contentInsetProp.bottom) != null ? _a : 0) + (horizontal ? 0 : keyboardInset.get()),
        left: (_b = contentInsetProp == null ? void 0 : contentInsetProp.left) != null ? _b : 0,
        right: (_c = contentInsetProp == null ? void 0 : contentInsetProp.right) != null ? _c : 0,
        top: (_d = contentInsetProp == null ? void 0 : contentInsetProp.top) != null ? _d : 0
      }
    }) : baseProps;
  });
  const style = isAndroid ? reactNativeReanimated.useAnimatedStyle(
    () => {
      var _a;
      return {
        ...styleFlattened || {},
        marginBottom: (_a = keyboardInset.get()) != null ? _a : 0
      };
    },
    [styleProp, keyboardInset]
  ) : void 0;
  return /* @__PURE__ */ React__namespace.createElement(
    reanimated.AnimatedLegendList,
    {
      ...rest,
      animatedProps,
      keyboardDismissMode: "interactive",
      onScroll: scrollHandler,
      ref: combinedRef,
      refScrollView: scrollViewRef,
      scrollIndicatorInsets: { bottom: 0, top: 0 },
      style
    }
  );
});

exports.KeyboardAvoidingLegendList = KeyboardAvoidingLegendList;
exports.LegendList = KeyboardAvoidingLegendList;
