// @legendapp/list/keyboard.mjs
import * as React from 'react';
import { forwardRef, useRef, useCallback } from 'react';
import { StyleSheet, Platform } from 'react-native';
import { useKeyboardHandler } from 'react-native-keyboard-controller';
import { useAnimatedRef, useSharedValue, useAnimatedScrollHandler, runOnJS, useAnimatedProps, useAnimatedStyle } from 'react-native-reanimated';
import { AnimatedLegendList } from '@legendapp/list/reanimated';

// src/integrations/keyboard.tsx

// src/utils/helpers.ts
function isFunction(obj) {
  return typeof obj === "function";
}

// src/hooks/useCombinedRef.ts
var useCombinedRef = (...refs) => {
  const callback = useCallback((element) => {
    for (const ref of refs) {
      if (!ref) {
        continue;
      }
      if (isFunction(ref)) {
        ref(element);
      } else {
        ref.current = element;
      }
    }
  }, refs);
  return callback;
};

// src/integrations/keyboard.tsx
var KeyboardAvoidingLegendList = forwardRef(function KeyboardAvoidingLegendList2(props, forwardedRef) {
  const {
    contentInset: contentInsetProp,
    horizontal,
    onScroll: onScrollProp,
    safeAreaInsetBottom = 0,
    style: styleProp,
    ...rest
  } = props;
  const styleFlattened = StyleSheet.flatten(styleProp);
  const refLegendList = useRef(null);
  const combinedRef = useCombinedRef(forwardedRef, refLegendList);
  const isIos = Platform.OS === "ios";
  const isAndroid = Platform.OS === "android";
  const scrollViewRef = useAnimatedRef();
  const scrollOffsetY = useSharedValue(0);
  const animatedOffsetY = useSharedValue(null);
  const scrollOffsetAtKeyboardStart = useSharedValue(0);
  const mode = useSharedValue("idle");
  const keyboardInset = useSharedValue(0);
  const keyboardHeight = useSharedValue(0);
  const isOpening = useSharedValue(false);
  const didInteractive = useSharedValue(false);
  const isKeyboardOpen = useSharedValue(false);
  const scrollHandler = useAnimatedScrollHandler(
    (event) => {
      scrollOffsetY.set(event.contentOffset[horizontal ? "x" : "y"]);
      if (onScrollProp) {
        runOnJS(onScrollProp)(event);
      }
    },
    [onScrollProp, horizontal]
  );
  const setScrollProcessingEnabled = useCallback(
    (enabled) => {
      var _a;
      (_a = refLegendList.current) == null ? void 0 : _a.setScrollProcessingEnabled(enabled);
    },
    [refLegendList]
  );
  useKeyboardHandler(
    // biome-ignore assist/source/useSortedKeys: prefer start/move/end
    {
      onStart: (event) => {
        "worklet";
        mode.set("running");
        if (isKeyboardOpen.get() && event.progress === 1 && event.height > 0) {
          return;
        }
        if (!didInteractive.get()) {
          if (event.height > 0) {
            keyboardHeight.set(event.height - safeAreaInsetBottom);
          }
          isOpening.set(event.progress > 0);
          scrollOffsetAtKeyboardStart.set(scrollOffsetY.get());
          animatedOffsetY.set(scrollOffsetY.get());
          runOnJS(setScrollProcessingEnabled)(false);
        }
      },
      onInteractive: (event) => {
        "worklet";
        if (mode.get() !== "running") {
          runOnJS(setScrollProcessingEnabled)(false);
        }
        mode.set("running");
        if (!didInteractive.get()) {
          didInteractive.set(true);
        }
        if (isAndroid && !horizontal) {
          keyboardInset.set(Math.max(0, event.height - safeAreaInsetBottom));
        }
      },
      onMove: (event) => {
        "worklet";
        if (!didInteractive.get()) {
          const vIsOpening = isOpening.get();
          const vKeyboardHeight = keyboardHeight.get();
          const vProgress = vIsOpening ? event.progress : 1 - event.progress;
          const targetOffset = Math.max(
            0,
            scrollOffsetAtKeyboardStart.get() + (vIsOpening ? vKeyboardHeight : -vKeyboardHeight) * vProgress
          );
          scrollOffsetY.set(targetOffset);
          animatedOffsetY.set(targetOffset);
          if (!horizontal) {
            keyboardInset.set(Math.max(0, event.height - safeAreaInsetBottom));
          }
        }
      },
      onEnd: (event) => {
        "worklet";
        const wasInteractive = didInteractive.get();
        const vMode = mode.get();
        mode.set("idle");
        if (vMode === "running") {
          if (!wasInteractive) {
            const vIsOpening = isOpening.get();
            const vKeyboardHeight = keyboardHeight.get();
            const targetOffset = Math.max(
              0,
              scrollOffsetAtKeyboardStart.get() + (vIsOpening ? vKeyboardHeight : -vKeyboardHeight) * (vIsOpening ? event.progress : 1 - event.progress)
            );
            scrollOffsetY.set(targetOffset);
            animatedOffsetY.set(targetOffset);
          }
          runOnJS(setScrollProcessingEnabled)(true);
          didInteractive.set(false);
          isKeyboardOpen.set(event.height > 0);
          if (!horizontal) {
            const newInset = Math.max(0, event.height - safeAreaInsetBottom);
            if (newInset > 0) {
              keyboardInset.set(newInset);
            } else {
              keyboardInset.set(newInset);
              animatedOffsetY.set(scrollOffsetY.get());
            }
          }
        }
      }
    },
    [scrollViewRef, safeAreaInsetBottom]
  );
  const animatedProps = useAnimatedProps(() => {
    "worklet";
    var _a, _b, _c, _d;
    const vAnimatedOffsetY = animatedOffsetY.get();
    const baseProps = {
      contentOffset: vAnimatedOffsetY === null ? void 0 : {
        x: 0,
        y: vAnimatedOffsetY
      }
    };
    return isIos ? Object.assign(baseProps, {
      contentInset: {
        bottom: ((_a = contentInsetProp == null ? void 0 : contentInsetProp.bottom) != null ? _a : 0) + (horizontal ? 0 : keyboardInset.get()),
        left: (_b = contentInsetProp == null ? void 0 : contentInsetProp.left) != null ? _b : 0,
        right: (_c = contentInsetProp == null ? void 0 : contentInsetProp.right) != null ? _c : 0,
        top: (_d = contentInsetProp == null ? void 0 : contentInsetProp.top) != null ? _d : 0
      }
    }) : baseProps;
  });
  const style = isAndroid ? useAnimatedStyle(
    () => {
      var _a;
      return {
        ...styleFlattened || {},
        marginBottom: (_a = keyboardInset.get()) != null ? _a : 0
      };
    },
    [styleProp, keyboardInset]
  ) : void 0;
  return /* @__PURE__ */ React.createElement(
    AnimatedLegendList,
    {
      ...rest,
      animatedProps,
      keyboardDismissMode: "interactive",
      onScroll: scrollHandler,
      ref: combinedRef,
      refScrollView: scrollViewRef,
      scrollIndicatorInsets: { bottom: 0, top: 0 },
      style
    }
  );
});

export { KeyboardAvoidingLegendList, KeyboardAvoidingLegendList as LegendList };
